### PR TITLE
make GCI image detection robust

### PR DESCRIPTION
**Pull request by adityakali**
This change makes sure that in case we roll back a released GCI image, the image detection logic picks a correct active image. 

@kubernetes/goog-image @Amey-D @wonderfly @dchen1107

### CODE REVIEW

File: cluster/gce/util.sh [hunk 1]
```diff
@@ -30,13 +30,30 @@ else
...
 if [[ "${OS_DISTRIBUTION}" == "gci" ]]; then
-  # If the master image is not set, we use the latest GCI dev image.
+  # If the master image is not set, we use the pinned GCI image.
   # Otherwise, we respect whatever is set by the user.
-  gci_images=( $(gcloud compute images list --project google-containers \
-      --show-deprecated --no-standard-images --sort-by='~creationTimestamp' \
-      --regexp='gci-[a-z]+-53-.*' --format='table[no-heading](name)') )
-  MASTER_IMAGE=${KUBE_GCE_MASTER_IMAGE:-"${gci_images[0]}"}
+  gci_image=$(get_latest_gci_image)
+  MASTER_IMAGE=${KUBE_GCE_MASTER_IMAGE:-"${gci_image}"}
```
Comment (roberthbailey) [inline cluster/gce/util.sh]: Can you collapse these two lines? I think doing so would skip the call to the get_latest_gci_image function in the case where the image is specified, which would be nice.

File: cluster/gce/util.sh [hunk 2]
```diff
@@ -30,13 +30,30 @@ else
   exit 1
 fi
 
+function get_latest_gci_image() {
+  # GCI milestone to use
+  milestone="53"
+
+  # First try to find an actie (non-deprecated) image on this milestone.
```
Comment (roberthbailey) [inline cluster/gce/util.sh]: active

File: cluster/gce/util.sh [hunk 3]
```diff
@@ -30,13 +30,30 @@ else
...
+      --regexp="gci-[a-z]+-${milestone}-.*" --format="table[no-heading](name)") )
+
+  # If no active image is available, search across all deprecated images.
+  if [[ ${#gci_images[@]} == 0 ]] ; then
+    gci_images=( $(gcloud compute images list --project google-containers \
+        --no-standard-images --show-deprecated --sort-by="~creationTimestamp" \
+        --regexp="gci-[a-z]+-${milestone}-.*" --format="table[no-heading](name)") )
+  fi
+
+  echo "${gci_images[0]}"
```
Comment (roberthbailey) [inline cluster/gce/util.sh]: should this be a return instead of an echo?

**Comment by roberthbailey** [comment]
Does this need to be in the 1.3 branch?

File: cluster/gce/util.sh (hunk 1 above)
Comment (adityakali) [inline cluster/gce/util.sh]: Good point. Done.

File: cluster/gce/util.sh (hunk 2 above)
Comment (adityakali) [inline cluster/gce/util.sh]: Done.

File: cluster/gce/util.sh (hunk 3 above)
Comment (adityakali) [inline cluster/gce/util.sh]: Humm.. I tried return "${gci_images[0]}" (and also without quotes), but I get error "numeric argument required". I couldn't find any other way to return a string from a bash function (but I am not a bash expert).

**Comment by adityakali** [comment]
`Does this need to be in the 1.3 branch?`

Yes please.

**Pull request merged**